### ISSUE TITLE
ZTunnel Safe Upgrade

### ISSUE THREAD

**Issue by therealmitchconnors** [istio/istio#51126]
ZTunnel runs as a daemonset, and current upgrade instructions qualify that upgrading the daemonset creates a short disruption to traffic.  

We should test and document a process for draining a node, and upgrading the ztunnel only on drained nodes, which prevents traffic disruptions for pods that gracefully drain their traffic on shutdown.  In-Place hitless upgrades of ztunnel, and multiple active ztunnel revisions per node are explicitly out of scope for this work.

**Comment by zirain** [comment]
if we have a large cluster(e.g. 100+ nodes, 2000+ pods), how would this upgrade happen?

**Comment by hzxuzhonghu** [comment]
Unless graceful connection migration, otherwise i cannot think of how can handle it

**Comment by bleggett** [comment]
> We should test and document a process for draining a node, and upgrading the ztunnel only on drained nodes, which prevents traffic disruptions for pods that gracefully drain their traffic on shutdown.

I'd expect this could potentially be achieved with taints / affinity and blue-green node pools: https://github.com/istio/istio/pull/50965

> We should test and document a process for draining a node, and upgrading the ztunnel only on drained nodes, which prevents traffic disruptions for pods that gracefully drain their traffic on shutdown.

How is this better than just the current inplace-upgrade? Seems both more expensive and more disruptive, with the same blast radius.

**Comment by therealmitchconnors** [comment]
Large clusters work the same as small clusters, with the exception that it would make sense to cordon, drain, and relabel groups of nodes all at once rather than one at a time.  

As for how draining is better than in-place upgrade, draining a node before upgrading the ztunnel ensures that no dataplane traffic is being handled by the node before replacing the ztunnel.  Effectively, the ztunnel will have no work to do while it is being upgraded.  This is the same process followed for upgrading the kubelet and other critical per-node resources.  Of course, not all users would choose to follow this guide, and can upgrade in place if they are comfortable with the risk of traffic disruption.

As for achieving the same outcome with affinity and taints, I would be interested in reviewing any concrete proposal, though I do not know of any way to do this myself.  Taints affect all pods scheduled on a node, meaning application pods would need to have tolerations added to them, which seems unnecessarily invasive.  Likewise, affinity is a sort of best-effort preference for scheduling, and I don't see how it could be useful for upgrading the ztunnel.  Let me know if I'm missing something here...

**Comment by bleggett** [comment]
> As for how draining is better than in-place upgrade, draining a node before upgrading the ztunnel ensures that no dataplane traffic is being handled by the node before replacing the ztunnel.  Effectively, the ztunnel will have no work to do while it is being upgraded. This is the same process followed for upgrading the kubelet and other critical per-node resources. Of course, not all users would choose to follow this guide, and can upgrade in place if they are comfortable with the risk of traffic disruption.

Right, my point is draining the entire node is a big traffic disruption event too - it's node-level traffic disruption either way, this is more hands-on if you want node-by-node control of the disruption events, I guess. 

Not against it, we need to document something, I think we should document both. Do you want a blip across all nodes, or do you want to manage per-node drains, operator choice.

> As for achieving the same outcome with affinity and taints, I would be interested in reviewing any concrete proposal, though I do not know of any way to do this myself. Taints affect all pods scheduled on a node, meaning application pods would need to have tolerations added to them, which seems unnecessarily invasive. Likewise, affinity is a sort of best-effort preference for scheduling, and I don't see how it could be useful for upgrading the ztunnel. Let me know if I'm missing something here...

I was thinking more taint the (new) nodes in the (new) pool (to block pod scheduling), add a toleration to the (new) ztunnel, put new ztunnel on new node with toleration, when ready remove taint (and wait for original pods to die a natural death- it's safe but leisurely)

**Comment by therealmitchconnors** [comment]
In theory node draining should not cause traffic disruption.  It sends a graceful shutdown signal to the pods, which should be able to conclude any pending requests, while new requests are routed to other pods.  In theory.

**Comment by hzxuzhonghu** [comment]
For tcp traffic, there is no way to notify server  or client to shutdown gracefully. ztunnel just pipe them, it doesnot even have any chance to notify both sides, it depend just on kernel notification

**Comment by howardjohn** [comment]
> For tcp traffic, there is no way to notify server or client to shutdown gracefully. ztunnel just pipe them, it doesnot even have any chance to notify both sides, it depend just on kernel notification

Yeah but a node drain can, since it sends SIGTERM to the app which they can handle

**Comment by bleggett** [comment]
> In theory node draining should not cause traffic disruption. It sends a graceful shutdown signal to the pods, which should be able to conclude any pending requests, while new requests are routed to other pods. In theory.

Yes, and in theory any app that can shutdown gracefully must also handle brief networking disruption, since networking disruptions are a fact of life.

But they don't.

There's not much point in making assumptions about what apps can or cannot handle.

All that matters is providing blast radius controls to the operator so they can take the model that they're comfortable with.

We have 2 - inplace and node drain - that should be enough to start with.

**Comment by zengyuxing007** [comment]
Operations personnel do not expect to see network issues in applications that are intentionally caused. It is common for application developers to challenge this by asking, "If I don't use a service mesh(ambient), won't I avoid these problems?"

Having a zero-downtime solution would be better. 

ref: https://github.com/istio/istio/issues/48949

**Comment by bleggett** [comment]
> Operations personnel do not expect to see network issues in applications that are intentionally caused. It is common for application developers to challenge this by asking, "If I don't use a service mesh(ambient), won't I avoid these problems?"

The answer to that would be "no, you **_will_** **_100_**% _**still**_ have these problems with no service mesh at all, and they will break your apps in exactly the same way".


> # The Eight Fallacies of Distributed Computing
> 
> ## Essentially everyone, when they first build a distributed application, makes the following eight assumptions. All prove to be false in the long run and all cause big trouble and painful learning experiences.
> 1. 	The network is reliable
> 2. 	Latency is zero
> 3. 	Bandwidth is infinite
> 4. 	The network is secure
> 5. 	Topology doesn't change
> 6. 	There is one administrator
> 7. 	Transport cost is zero
> 8. 	The network is homogeneous

The network is _not_ reliable or zero-downtime in K8S, and if your app assumes it is, your app will break in kubernetes, with or without a service mesh.

**Comment by d-mankowski-synerise** [comment]
> I'd expect this could potentially be achieved with taints / affinity and blue-green node pools: https://github.com/istio/istio/pull/50965

We went with `OnDelete` [update strategy](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy) - TLDR - we simply bump image version, which **does not** restart/change existing ztunnel pods. In case of spot instances, we simply wait for all of those to be evicted/recreated (new nodes do use new ztunnel version), in case of regular node pools - we cordon all of existing nodes, and then drain them.

In case of `taints / affinity and blue-green node pools`, documentation should explicitly mention, that in istiod it is required to set `CA_TRUSTED_NODE_ACCOUNTS` with all `{namespace}/{service account}` of all ztunnel daemonsets, since current implementation of helm chart always creates (and uses) separate service account per daemonset (this can be mitigated by e.g. patching daemonsets with kustomize, to use the same service account by all daemonsets).

> the node drain inherently restarts workloads...

If application is not able to handle restart, then, in general, it should not be deployed on Kubernetes. Every Kubernetes update causes node drain.

**Comment by Unichron** [comment]
> We went with `OnDelete` [update strategy](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy) - TLDR - we simply bump image version, which **does not** restart/change existing ztunnel pods. In case of spot instances, we simply wait for all of those to be evicted/recreated (new nodes do use new ztunnel version), in case of regular node pools - we cordon all of existing nodes, and then drain them.

This sounds like a great tip to try out!

> If application is not able to handle restart, then, in general, it should not be deployed on Kubernetes. Every Kubernetes update causes node drain.

I didn't mean that restarting workloads is not zero-downtime. I meant that the premise of ambient about not needing to restart workloads is not true, if one want zero-downtime. So, you either restart workloads or you face downtime risk.
